File: src/main/java/com/matag/cards/CardUtils.java
package com.matag.cards;

import com.matag.cards.properties.Color;
import com.matag.cards.properties.Type;

import java.util.Set;

public class CardUtils {
  public static boolean isColorless(Card card) {
    return card.getColors().isEmpty();
  }

  public static boolean isMulticolor(Card card) {
    return card.getColors().size() > 1;
  }

  public static boolean isOfColor(Card card, Color color) {
    return card.getColors().contains(color);
  }

  public static boolean isOfOnlyAnyOfTheColors(Card card, Set<Color> colors) {
    if (card.getColors().isEmpty()) {
      return false;
    }

    for (Color color : card.getColors()) {
      if (!colors.contains(color)) {
        return false;
      }
    }
    return true;
  }

  public static boolean isNotOfType(Card card, Type type) {
    return !isOfType(card, type);
  }

  public static boolean isOfType(Card card, Type type) {
    return card.getTypes().contains(type);
  }
}

File: src/main/kotlin/com/matag/cards/CardUtils.kt
package com.matag.cards

import com.matag.cards.properties.Color
import com.matag.cards.properties.Subtype
import com.matag.cards.properties.Type

fun Card.isOfType(type: Type) =
    this.types?.contains(type) ?: false

fun Card.isNotOfType(type: Type) =
    !this.isOfType(type)

fun Card.isOfSubtype(subtype: Subtype) =
    this.subtypes?.contains(subtype) ?: false

fun Card.isNotOfSubtype(subtype: Subtype) =
    !this.isOfSubtype(subtype)

fun Card.isColorless(): Boolean {
    return this.colors().isEmpty()
}

fun Card.isMulticolor(): Boolean {
    return this.colors().size > 1
}

fun Card.isOfColor(color: Color?): Boolean {
    return this.colors().contains(color)
}

fun Card.isOfOnlyAnyOfTheColors(colors: Set<Color>): Boolean {
    if (this.colors().isEmpty()) {
        return false
    }

    for (color in this.colors()) {
        if (!colors.contains(color)) {
            return false
        }
    }
    return true
}
